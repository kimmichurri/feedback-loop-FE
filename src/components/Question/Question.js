import React, { Component } from 'react'

export class Question extends Component {
  constructor(props) {
    super(props);
    this.state = {
      id: this.props.id,
      questionTitle: '',
      option_1: {pointValue: 4, description: ''},
      option_2: {pointValue: 3, description: ''},
      option_3: {pointValue: 2, description: ''},
      option_4: {pointValue: 1, description: ''}
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target
      this.setState({  [name] : {pointValue: this.state[name].pointValue, description: value} })
  }

  handleTitleChange = (e) => {
    this.setState({
      questionTitle: e.target.value
    })
  }

  handleSubmit = (e) => {
    e.preventDefault()
    let optionsArr = []
    Object.keys(this.state).forEach(el => {
      if(el.includes('option')) {
        optionsArr.push({ [el] : {pointValue: this.state[el].pointValue, description: this.state[el].description}})
      }
    })
    let newQuestion = {
      id: this.state.id,
      questionTitle: this.state.questionTitle,
      options: optionsArr
    }
    this.props.updateQuestions(newQuestion)
  }

  render() {
    return (
      <form onBlur={this.handleSubmit} className='question-form'>
        <div>
          <input
            type="text"
            name="questionTitle"
            className="question-input"
            placeholder="Add Your Question Text"
            value={this.state.questionTitle}
            onChange={this.handleTitleChange}
            autoComplete="off"
          />
        </div>
        <div className='option-container'>
          <div className='option'>
            <input
              type="radio"
              name="radio" />
            <p>4.</p>
            <input
              type="text"
              placeholder="Add Response Description"
              name="option_1"
              id='op1'
              className='option'
              value={this.state.option_1.description}
              onChange={this.handleChange} 
              autoComplete="off"
            />
          </div>
          <div className='option'>
            <input
              type="radio"
              name="radio" />
            <p>3.</p>
            <input
              type="text"
              placeholder="Add Response Description"
              name="option_2"
              className='option'
              value={this.state.option_2.description}
              onChange={this.handleChange} 
              autoComplete="off"
            />
          </div>
          <div className='option'>
            <input
              type="radio"
              name="radio" />
            <p>2.</p>
            <input
              type="text"
              placeholder="Add Response Description"
              name="option_3"
              className='option'
              value={this.state.option_3.description}
              onChange={this.handleChange} 
              autoComplete="off"  
            />
          </div>
          <div className='option'>
            <input
              type="radio"
              name="radio" />
            <p>1.</p>
            <input
              type="text"
              placeholder="Add Response Description"
              name="option_4"
              className='option'
              value={this.state.option_4.description}
              onChange={this.handleChange} 
              autoComplete="off"  
            />
          </div>
        </div>
      </form>
    )
  }
}

export default Question