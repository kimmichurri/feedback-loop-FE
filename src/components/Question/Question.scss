.question-form {
  display: flex;
  flex-direction: column;
  text-align: left;
  padding: 30px 10px;
  background: $white;
  border: solid #f4f4f4 .5px;
  box-shadow: 0 3px 5px rgba(0,0,0,0.20);
  margin-bottom: 5px;
  margin-top: 5px;
}

.question-input {
  width: 100%;
  border: none;
  border-bottom: solid $main-grey 1px;
  font-size: 20px;
  color: $accent-charcoal;
}

.option-container {
  display: flex;
  flex-direction: column;
}

.option {
  width: 95%;
  font-size: 13px;
  height: 50px;
  padding-left: 5px;
  border: none;
  color: $accent-charcoal;
    &:focus {
      outline: none;
    }
}
