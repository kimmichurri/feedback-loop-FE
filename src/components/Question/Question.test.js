import React from 'react'
import { Question } from './Question'
import { shallow } from 'enzyme'

describe('Question', () => {
  let wrapper
  let mockFn = jest.fn()

  beforeEach(() => {
    wrapper = shallow(<Question 
      id='abc'
      updateQuestions={mockFn}
    />)
  })
  
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have a default state', () => {
    const defaultState = {
      id: 'abc',
      questionTitle: '',
      option_1: {pointValue: 1, questionText: ''},
      option_2: {pointValue: 2, questionText: ''},
      option_3: {pointValue: 3, questionText: ''},
      option_4: {pointValue: 4, questionText: ''},
      option_5: {pointValue: 5, questionText: ''}
    }

    expect(wrapper.state()).toEqual(defaultState)
  })

  it('should update questionText in state on change', () => {
    const mockEvent = { target: { name: 'option_1', value: 'First Option'}}
    wrapper.find('#op1').simulate('change', mockEvent)
    expect(wrapper.state('option_1')).toEqual({ "pointValue": 1, "questionText": 'First Option'})
  })

  it('should update questionTitle in state on change', () => {
    const mockEvent = { target: { name: 'questionTitle', value: 'My Question'} }
    wrapper.find('.question-input').simulate('change', mockEvent)
    expect(wrapper.state('questionTitle')).toBe('My Question')
  })

  it('should call updateQuestions when handleSubmit is invoked', () => {
    const preventDefault = { preventDefault: jest.fn()}
    const updatedState = {
      id: 'abc',
      questionTitle: 'New Q',
      option_1: {pointValue: 1, questionText: 'a'},
      option_2: {pointValue: 2, questionText: 'b'},
      option_3: {pointValue: 3, questionText: 'c'},
      option_4: {pointValue: 4, questionText: 'd'},
      option_5: {pointValue: 5, questionText: 'e'}
    }
    
    wrapper.setState(updatedState)
    wrapper.instance().handleSubmit(preventDefault)
    expect(mockFn).toHaveBeenCalledWith(updatedState)
  })
})
