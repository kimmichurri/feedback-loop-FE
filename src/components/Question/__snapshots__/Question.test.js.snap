// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question should match the snapshot 1`] = `
<form
  className="question-form"
  onBlur={[Function]}
>
  <div>
    <input
      className="question-input"
      name="questionTitle"
      onChange={[Function]}
      placeholder="Untitled Question"
      type="text"
      value=""
    />
  </div>
  <div
    className="option-container"
  >
    <div
      className="option"
    >
      <input
        name="radio"
        type="radio"
      />
      <input
        className="option"
        id="op1"
        name="option_1"
        onChange={[Function]}
        placeholder="Option description"
        type="text"
        value=""
      />
    </div>
    <div
      className="option"
    >
      <input
        name="radio"
        type="radio"
      />
      <input
        className="option"
        name="option_2"
        onChange={[Function]}
        placeholder="Option description"
        type="text"
        value=""
      />
    </div>
    <div
      className="option"
    >
      <input
        name="radio"
        type="radio"
      />
      <input
        className="option"
        name="option_3"
        onChange={[Function]}
        placeholder="Option description"
        type="text"
        value=""
      />
    </div>
    <div
      className="option"
    >
      <input
        name="radio"
        type="radio"
      />
      <input
        className="option"
        name="option_4"
        onChange={[Function]}
        placeholder="Option description"
        type="text"
        value=""
      />
    </div>
    <div
      className="option"
    >
      <input
        name="radio"
        type="radio"
      />
      <input
        className="option"
        name="option_5"
        onChange={[Function]}
        placeholder="Option description"
        type="text"
        value=""
      />
    </div>
  </div>
</form>
`;
